storage/remote/storage.go

```diff
@@ -96,10 +98,21 @@ func (s *Storage) ApplyConfig(conf *config.Config) error {
 	s.configHash = hash
 
 	// Update write queues
-	newQueues := []*QueueManager{}
-	// TODO: we should only stop & recreate queues which have changes,
-	// as this can be quite disruptive.
+	unseen := make(map[string]*QueueManager)
+	common := make(map[string]struct{})
 	for i, rwConf := range conf.RemoteWriteConfigs {
+		b, err := json.Marshal(rwConf)
+		if err != nil {
+			return err
+		}
+
+		hash := fmt.Sprintf("%x", md5.Sum(b))
+		if _, ok := s.queues[hash]; ok {
+			// Skip unchanged queue
+			common[hash] = struct{}{}
+			continue
+		}
+
 		c, err := NewClient(i, &ClientConfig{
```

Comment:
We pass `i` to the client and use that for at least metrics. I think after this change you could end up in a state where multiple clients have the same identifier.